Cabal/src/Distribution/Simple/Program/Ld.hs

```diff
@@ -104,3 +106,10 @@ combineObjectFiles verbosity lbi ld target files = do
       runProgramInvocation verbosity inv
       renameFile target tmpfile
       run invs
+
+    -- Prepend "-r" to the list if the linker supports relocatable outputs.
+    prependRelocatableFlag :: [String] -> [String]
```

Comment:
I'm a bit lost here. It seems to me that this will simply attempt to call `ld` without `-r` to join objects when `-r` is not supported. This is, AFAICT, not correct and will at best result in linking (undefined reference) errors, and at worst result in an invalid (non-relocatable) object.

Reply:
If `ldProgram` doesn’t have `-r`, we shouldn’t even try to merge objects.

Reply:
Updated to call `error` if `combineObjectFiles` is called with an `ldProgram` that does not support relocatable output.

Reply:
@Mikolaj is `error` the best way to fail here?

Reply:
I had a bit of a look, but `error` is used quite extensively elsewhere in the `Cabal` library.

Reply:
I believe `dieWithException` is the standard after #9219